<a href="https://colab.research.google.com/github/Narsimha1729/comply/blob/main/Comply.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Basic working of Google Palm LLM in LangChain

In [35]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [36]:
!pip install jupyter-dash

In [37]:
import plotly.express as px
from jupyter_dash import JupyterDash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output# Load Data

In [38]:
pip install -r /content/drive/MyDrive/3_project_codebasics_q_and_a/requirements.txt

  Using cached langchain-0.0.284-py3-none-any.whl (1.7 MB)
  Using cached langsmith-0.0.92-py3-none-any.whl (56 kB)
  Attempting uninstall: langsmith
    Found existing installation: langsmith 0.1.62
    Uninstalling langsmith-0.1.62:
      Successfully uninstalled langsmith-0.1.62
  Attempting uninstall: langchain
    Found existing installation: langchain 0.2.0
    Uninstalling langchain-0.2.0:
      Successfully uninstalled langchain-0.2.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
langchain-community 0.2.0 requires langchain<0.3.0,>=0.2.0, but you have langchain 0.0.284 which is incompatible.
langchain-community 0.2.0 requires langsmith<0.2.0,>=0.1.0, but you have langsmith 0.0.92 which is incompatible.
langchain-core 0.2.1 requires langsmith<0.2.0,>=0.1.0, but you have langsmith 0.0.92 which is incompatible.


In [39]:
!pip install --upgrade langchain
!pip install --upgrade langchain langchain-community


  Using cached langchain-0.2.0-py3-none-any.whl (973 kB)
  Using cached langsmith-0.1.62-py3-none-any.whl (122 kB)
  Attempting uninstall: langsmith
    Found existing installation: langsmith 0.0.92
    Uninstalling langsmith-0.0.92:
      Successfully uninstalled langsmith-0.0.92
  Attempting uninstall: langchain
    Found existing installation: langchain 0.0.284
    Uninstalling langchain-0.0.284:
      Successfully uninstalled langchain-0.0.284


In [40]:
!pip install google-generativeai

In [41]:
# !pip install --upgrade langchain
# from langchain.llms import GooglePalm

# api_key = 'AIzaSyA6WQjKyCPsqkuI2pmTGThK7LWgvGdsbQE'

# llm = GooglePalm(google_api_key=api_key, temperature=0.1)
from langchain_community.llms import GooglePalm

api_key = 'AIzaSyA6WQjKyCPsqkuI2pmTGThK7LWgvGdsbQE'  # Replace with your actual API key

llm = GooglePalm(google_api_key=api_key, temperature=0.1)

# Example usage
response = llm("Tell me a joke.")
print(response)


Why did the scarecrow win an award? Because he was outstanding in his field!


In [42]:
poem = llm("Write a 4 line poem of my love for AI")
print(poem)

AI, my love, you are so smart,
You learn and grow with each passing day.
I can't wait to see what the future holds
For you and me, together.


In [43]:
essay = llm("write email requesting refund for FLIPKART buy")
print(essay)

Dear Flipkart Customer Service,

I am writing to request a refund for the item I purchased on [date]. The item number is [item number] and the cost was [cost].

I am requesting a refund because the item was not as described. The product description said that the item was "brand new" but when I received it, it was clearly used. The item was also damaged.

I have attached photos of the item to this email.

I would like to request a full refund for the item. I have already returned the item to Flipkart.

I would appreciate it if you could process my refund as soon as possible.

Thank you for your time and attention to this matter.

Sincerely,
[Your name]


In [44]:
from langchain.chains import RetrievalQA


from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

### Now let's load data from Codebasics FAQ csv file

In [45]:
from langchain.document_loaders.csv_loader import CSVLoader

# Initialize the CSVLoader with the correct file path and column name
loader = CSVLoader(file_path='/content/drive/MyDrive/vaid/jsa-data-in-excel.csv', source_column="prompt", encoding="latin-1")

# Load the data
data = loader.load()



### Hugging Face Embeddings

In [46]:
!pip install sentence-transformers==2.2.2

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 836.1 kB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.20.5-py3-none-manylinux2014_x86_64.whl (176.2 MB)
  Using ca

In [47]:
pip install InstructorEmbedding

In [48]:
from langchain.embeddings import HuggingFaceInstructEmbeddings

# Initialize instructor embeddings using the Hugging Face model
instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")

e = instructor_embeddings.embed_query("What is your  policy?")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


.gitattributes:   0%|          | 0.00/1.48k [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/270 [00:00<?, ?B/s]

2_Dense/config.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/3.15M [00:00<?, ?B/s]

README.md:   0%|          | 0.00/66.3k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.53k [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.34G [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.41k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/461 [00:00<?, ?B/s]

load INSTRUCTOR_Transformer
max_seq_length  512


In [49]:
len(e)

768

In [50]:
e[:5]

[-0.016029156744480133,
 -0.016262713819742203,
 -0.03525111451745033,
 0.026284627616405487,
 0.025402333587408066]

As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

### Vector store using FAISS

In [51]:
from langchain.vectorstores import FAISS

# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=data,
                                 embedding=instructor_embeddings)

# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)

In [52]:
rdocs = retriever.get_relevant_documents("how about job placement support?")
rdocs

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:119: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 0.3.0. Use invoke instead.
  warn_deprecated(


[Document(page_content='prompt: \nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.', metadata={'source': '', 'row': 11}),
 Document(page_content='prompt: \nresponse: The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.', metadata={'source': '', 'row': 15}),
 Document(page_content='prompt: \nresponse: Yes, this bootcamp will certainly help because we cover the majority of the skills measured in these exams. However, please be informed that this course focuses on Job ready aspects and not on all aspects

As you can see above, the retriever that was created using FAISS and hugging face embedding is now capable of pulling relavant documents from our original CSV file knowledge store. This is very powerful and it will help us further in our project

##### Embeddings can be created using GooglePalm too. Also for vector database you can use chromadb as well as shown below. During our experimentation, we found hugging face embeddings and FAISS to be more appropriate for our use case

In [53]:
# google_palm_embeddings = GooglePalmEmbeddings(google_api_key=api_key)

# from langchain.vectorstores import Chroma
# vectordb = Chroma.from_documents(data,
#                            embedding=google_palm_embeddings,
#                            persist_directory='./chromadb')
# vectordb.persist()

### Create RetrievalQA chain along with prompt template 🚀

In [54]:
from langchain.prompts import PromptTemplate

prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}


from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs=chain_type_kwargs)


### We are all set 👍🏼 Let's ask some questions now

import plotly.express as px
from jupyter_dash import JupyterDash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output# Load Data

In [55]:
chain('Do you provide job assistance and also do you provide job gurantee?')

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:119: LangChainDeprecationWarning: The method `Chain.__call__` was deprecated in langchain 0.1.0 and will be removed in 0.3.0. Use invoke instead.
  warn_deprecated(


{'query': 'Do you provide job assistance and also do you provide job gurantee?',
 'result': 'Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters. The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.',
 'source_documents': [Document(page_content='prompt: \nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.', metadata={'source': '', 'row': 11}),
  Document(page_content='prom

**As you can see above, the answer of question comes from two different FAQs within our csv file and it is able to pull those questions and merge them nicely**

In [56]:
chain("Do you guys provide internship and also do you offer EMI payments?")

{'query': 'Do you guys provide internship and also do you offer EMI payments?',
 'result': 'Yes, we provide internship and also offer EMI payments.',
 'source_documents': [Document(page_content='prompt: \nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.', metadata={'source': '', 'row': 11}),
  Document(page_content='prompt: \nresponse: The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.', metadata={'source': '', 'row': 15}),
  Document(page_content='prompt: Abbreviation\nresponse: We have created eve

In [57]:
chain("do you have  course?")

{'query': 'do you have  course?',
 'result': 'Yes, we have a course on Excel.',
 'source_documents': [Document(page_content='prompt: \nresponse: The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.', metadata={'source': '', 'row': 15}),
  Document(page_content='prompt: \nresponse: Yes, this is the perfect course for anyone who has never worked on excel and wants to build a career in the IT/Data Analytics industry or just wants to perform better in their current job or business using data.', metadata={'source': '', 'row': 16}),
  Document(page_content='prompt: Annexures\nresponse: Yes. This bootcamp is self-paced. You can learn on your own schedu

In [58]:
chain("Do you have plans to launch blockchain course in future?")

{'query': 'Do you have plans to launch blockchain course in future?',
 'result': 'Yes, we are planning to launch a blockchain course in the future.',
 'source_documents': [Document(page_content='prompt: \nresponse: The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.', metadata={'source': '', 'row': 15}),
  Document(page_content='prompt: Abbreviation\nresponse: We have created every lecture with a motive to explain everything in an easy-to-understand manner. While working on these lectures you could make mistakes in steps or have some doubts. You need to commit yourself to hold patience, make efforts & diagnose the errors yourself by googling in

In [59]:
chain("should I learn power bi or tableau?")

{'query': 'should I learn power bi or tableau?',
 'result': "I don't know.",
 'source_documents': [Document(page_content='prompt: Revision History\nresponse: You can expect to learn tech skills like Python, Excel, SQL, and Power BI along with other skills like basic project management, stakeholder management, communication skills, and Business knowledge for domains like Finance, Supply chain, Sales, and Marketing.', metadata={'source': 'Revision History', 'row': 10}),
  Document(page_content='prompt: \nresponse: Yes, this is the perfect course for anyone who has never worked on excel and wants to build a career in the IT/Data Analytics industry or just wants to perform better in their current job or business using data.', metadata={'source': '', 'row': 16}),
  Document(page_content='prompt: \nresponse: Yes, this bootcamp will certainly help because we cover the majority of the skills measured in these exams. However, please be informed that this course focuses on Job ready aspects and 

In [60]:
chain("I've a MAC computer. Can I use powerbi on it?")

{'query': "I've a MAC computer. Can I use powerbi on it?",
 'result': 'Yes, you can use Power BI on a Mac',
 'source_documents': [Document(page_content='prompt: \nresponse: Yes, this bootcamp will certainly help because we cover the majority of the skills measured in these exams. However, please be informed that this course focuses on Job ready aspects and not on all aspects required to clear the exams. In addition to this course, you might need to visit the official learning material designed by Microsoft which is available for free on their website.', metadata={'source': '', 'row': 12}),
  Document(page_content='prompt: Annexures\nresponse: Yes. This bootcamp is self-paced. You can learn on your own schedule.', metadata={'source': 'Annexures', 'row': 9}),
  Document(page_content='prompt: \nresponse: Yes', metadata={'source': '', 'row': 14}),
  Document(page_content='prompt: Authorization of the permit\nresponse: Yes', metadata={'source': 'Authorization of the permit', 'row': 7})]}

In [61]:
chain("I don't see power pivot. how can I enable it?")

{'query': "I don't see power pivot. how can I enable it?",
 'result': 'To enable Power Pivot, go to File > Options > Add-ins > Manage: COM Add-ins > Go. In the Add-ins dialog box, select Microsoft Power Pivot for Excel and click OK.',
 'source_documents': [Document(page_content='prompt: Revision History\nresponse: You can expect to learn tech skills like Python, Excel, SQL, and Power BI along with other skills like basic project management, stakeholder management, communication skills, and Business knowledge for domains like Finance, Supply chain, Sales, and Marketing.', metadata={'source': 'Revision History', 'row': 10}),
  Document(page_content='prompt: \nresponse: Yes, this is the perfect course for anyone who has never worked on excel and wants to build a career in the IT/Data Analytics industry or just wants to perform better in their current job or business using data.', metadata={'source': '', 'row': 16}),
  Document(page_content='prompt: \nresponse: No', metadata={'source': '',

In [62]:
 chain("What is the price of your machine learning course?")

{'query': 'What is the price of your machine learning course?',
 'result': '19999',
 'source_documents': [Document(page_content='prompt: \nresponse: The courses included in this bootcamp are done by 9000+ learners and many of them have secured a job which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.', metadata={'source': '', 'row': 15}),
  Document(page_content='prompt: Abbreviation\nresponse: We have created every lecture with a motive to explain everything in an easy-to-understand manner. While working on these lectures you could make mistakes in steps or have some doubts. You need to commit yourself to hold patience, make efforts & diagnose the errors yourself by googling in order to become truly job ready. For any questions, that Google c

In [63]:
 chain("objective for chiller installation sop")

{'query': 'objective for chiller installation sop',
 'result': 'This SOP shall be applicable for Chillers 100 TR',
 'source_documents': [Document(page_content='prompt: Objective\nresponse: To provide procedure for operation and maintenance of portable fire extinguishers.\n\nTo describe a procedure for the operation of air handling units.\n\nTo lay down a procedure for bubble point test of sterilizing grade filter.\n\nThe purpose of this SOP is to provide stepwise instructions for Building Maintenance.\n\nTo provide guidelines for the operation of the effluent treatment plant\n\nThe purpose of this SOP is to provide stepwise instructions for Preventive Maintenance of AHU & Ventilation units.\n\nThe purpose of this SOP is to provide stepwise instructions for planned preventive maintenance programme\n\nTo lay down the procedure for the intimation of the breakdown of utilities services & equipment in manufacturing & packing department during working.\n\nTo lay down the procedure for to pro